tests/nvapi_d3d12.cpp

```diff
@@ -15,4 +15,165 @@ TEST_CASE("D3D12 methods succeed", "[.d3d12]") {
+        D3D12_CPU_DESCRIPTOR_HANDLE srvHandle;
+
+        REQUIRE(NvAPI_D3D12_GetCudaTextureObject(static_cast<ID3D12Device*>(&device), srvHandle, srvHandle, nullptr)  == NVAPI_ERROR);
+        REQUIRE(NvAPI_D3D12_GetCudaSurfaceObject(static_cast<ID3D12Device*>(&device), srvHandle, nullptr)  == NVAPI_ERROR);
+
+        REQUIRE(NvAPI_D3D12_CaptureUAVInfo(static_cast<ID3D12Device*>(&device), nullptr) == NVAPI_ERROR);
+        
+        bool isPTXSupported;
+        REQUIRE(NvAPI_D3D12_IsFatbinPTXSupported(static_cast<ID3D12Device*>(&device), &isPTXSupported) == NVAPI_ERROR);
+    }
+
+    SECTION("NvAPI_D3D12_GetGraphicsCapabilities returns ok") {
+        NV_D3D12_GRAPHICS_CAPS graphicsCaps;
+        graphicsCaps.bFastUAVClearSupported = false;
+        REQUIRE(NvAPI_D3D12_GetGraphicsCapabilities(static_cast<ID3D12Device*>(&device), NV_D3D12_GRAPHICS_CAPS_VER1, &graphicsCaps) == NVAPI_OK);
+        REQUIRE(graphicsCaps.bFastUAVClearSupported == true);
+    }
+
+    SECTION("D3D12 methods with vkd3d-proton returns OK ") {
+        REQUIRE_CALL(device, QueryInterface(ID3D12DeviceExt::guid, _))
+                .LR_SIDE_EFFECT(*_2 = static_cast<ID3D12DeviceExt*>(&deviceExt))
+                .RETURN(S_OK)
+                .TIMES(1);
+
+        ALLOW_CALL(deviceExt, GetExtensionSupport(D3D12_VK_NVX_BINARY_IMPORT))
+            .RETURN(true);
+        
+        const void*                   cubin_data = nullptr;
+        UINT32                        cubin_size = 2;
+        UINT32                        block_x = 3;
+        UINT32                        block_y = 4;
+        UINT32                        block_z = 5;
+        const char*                   shader_name = "";
+        D3D12_CUBIN_DATA_HANDLE*      shader_handle = reinterpret_cast<D3D12_CUBIN_DATA_HANDLE*>(0x912122);
+        D3D12_CUBIN_DATA_HANDLE**     handle = &shader_handle;
+
+        ALLOW_CALL(deviceExt, CreateCubinComputeShaderWithName(cubin_data, cubin_size, block_x, block_y, block_z, shader_name, handle))
+            .RETURN(S_OK);
+        REQUIRE(NvAPI_D3D12_CreateCubinComputeShaderWithName(static_cast<ID3D12Device*>(&device), cubin_data, cubin_size, block_x, block_y, block_z, shader_name, reinterpret_cast<NVDX_ObjectHandle*>(handle)) == NVAPI_OK);
+        REQUIRE(NvAPI_D3D12_CreateCubinComputeShader(static_cast<ID3D12Device*>(&device), cubin_data, cubin_size, block_x, block_y, block_z, reinterpret_cast<NVDX_ObjectHandle*>(handle)) == NVAPI_OK);
+    
+    
+        ALLOW_CALL(deviceExt, DestroyCubinComputeShader(shader_handle))
+            .RETURN(S_OK);
+        REQUIRE(NvAPI_D3D12_DestroyCubinComputeShader(static_cast<ID3D12Device*>(&device), reinterpret_cast<NVDX_ObjectHandle>(shader_handle)) == NVAPI_OK);
+
+       /*  
+           TODO : Uncomment this once compilation issues are resolved for below calls. 
+        D3D12_CPU_DESCRIPTOR_HANDLE srvHandle = {0}, samplerHandle = {0};   
+        ALLOW_CALL(deviceExt, GetCudaTextureObject(srvHandle, srvHandle, nullptr))
```

Comment:
@jp7677 Could you please help me to solve this issue here. `ALLOW_CALL`  throws an error here. However, `deviceExt.GetCudaTextureObject() `  works fine.

Reply:
I'm currently at work, will take a look in the evening (CEST here).
Does it work with `_` ?

Reply:
`In file included from ../../tests/../inc/catch2/trompeloeil.hpp:23,
                 from ../../tests/nvapi_d3d12.cpp:4:
../../tests/../inc/catch2/../trompeloeil.hpp: In instantiation of ‘bool trompeloeil::param_matches_impl(const T&, const U&, ...) [with T = D3D12_CPU_DESCRIPTOR_HANDLE; U = D3D12_CPU_DESCRIPTOR_HANDLE]’:
../../tests/../inc/catch2/../trompeloeil.hpp:3151:39:   required from ‘bool trompeloeil::param_matches(const T&, const U&) [with T = D3D12_CPU_DESCRIPTOR_HANDLE; U = D3D12_CPU_DESCRIPTOR_HANDLE]’
../../tests/../inc/catch2/../trompeloeil.hpp:3162:77:   required from ‘bool trompeloeil::match_parameters(std::index_sequence<I ...>, const T&, const U&) [with long long unsigned int ...I = {0, 1}; T = std::tuple<D3D12_CPU_DESCRIPTOR_HANDLE, std::nullptr_t>; U = std::tuple<D3D12_CPU_DESCRIPTOR_HANDLE&, unsigned int*&>; std::index_sequence<I ...> = std::integer_sequence<long long unsigned int, 0, 1>]’
../../tests/../inc/catch2/../trompeloeil.hpp:3175:37:   required from ‘bool trompeloeil::match_parameters(const std::tuple<_Tps ...>&, const std::tuple<_Elements ...>&) [with T = {D3D12_CPU_DESCRIPTOR_HANDLE, std::nullptr_t}; U = {D3D12_CPU_DESCRIPTOR_HANDLE&, unsigned int*&}]’
../../tests/../inc/catch2/../trompeloeil.hpp:3962:30:   required from ‘bool trompeloeil::call_matcher<Sig, Value>::matches(trompeloeil::call_params_type_t<Sig>&) const [with Sig = long int(D3D12_CPU_DESCRIPTOR_HANDLE, unsigned int*); Value = std::tuple<D3D12_CPU_DESCRIPTOR_HANDLE, std::nullptr_t>; trompeloeil::call_params_type_t<Sig> = std::tuple<D3D12_CPU_DESCRIPTOR_HANDLE&, unsigned int*&>]’
../../tests/../inc/catch2/../trompeloeil.hpp:3957:5:   required from here
../../tests/../inc/catch2/../trompeloeil.hpp:3141:51: error: no match for ‘operator==’ (operand types are ‘D3D12_CPU_DESCRIPTOR_HANDLE’ and ‘const D3D12_CPU_DESCRIPTOR_HANDLE’)
 3141 |   noexcept(noexcept(::trompeloeil::identity<U>(t) == u))
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~
In file included from /usr/share/mingw-w64/include/winnt.h:635,
                 from /usr/share/mingw-w64/include/minwindef.h:163,
                 from /usr/share/mingw-w64/include/windef.h:9,
                 from /usr/share/mingw-w64/include/windows.h:69,
                 from /usr/share/mingw-w64/include/rpc.h:16,
                 from /usr/share/mingw-w64/include/dxgi.h:7,
                 from ../../tests/../src/nvapi_private.h:15,
                 from ../../tests/../src/nvapi_d3d12.cpp:1,
                 from ../../tests/nvapi_d3d12.cpp:5:
/usr/share/mingw-w64/include/guiddef.h:180:15: note: candidate: ‘bool operator==(const GUID&, const GUID&)’
  180 | __inline bool operator== (REFGUID guidOne, REFGUID guidOther) { return !!IsEqualGUID (guidOne, guidOther); }
      |               ^~~~~~~~
/usr/share/mingw-w64/include/guiddef.h:180:35: note:   no known conversion for argument 1 from ‘D3D12_CPU_DESCRIPTOR_HANDLE’ to ‘const GUID&’ {aka ‘const _GUID&’}
  180 | __inline bool operator== (REFGUID guidOne, REFGUID guidOther) { return !!IsEqualGUID (guidOne, guidOther); }
      |                                   ^
In file included from ../../tests/../inc/catch2/trompeloeil.hpp:23,
                 from ../../tests/nvapi_d3d12.cpp:4:
../../tests/../inc/catch2/../trompeloeil.hpp: In instantiation of ‘bool trompeloeil::param_matches_impl(const T&, const U&, ...) [with T = D3D12_CPU_DESCRIPTOR_HANDLE; U = D3D12_CPU_DESCRIPTOR_HANDLE]’:
../../tests/../inc/catch2/../trompeloeil.hpp:3153:45:   required from ‘bool trompeloeil::param_matches(const T&, const U&) [with T = D3D12_CPU_DESCRIPTOR_HANDLE; U = D3D12_CPU_DESCRIPTOR_HANDLE]’
../../tests/../inc/catch2/../trompeloeil.hpp:3187:38:   required from ‘void trompeloeil::print_mismatch(std::ostream&, size_t, const V&, const P&) [with V = D3D12_CPU_DESCRIPTOR_HANDLE; P = D3D12_CPU_DESCRIPTOR_HANDLE; std::ostream = std::basic_ostream<char>; size_t = long long unsigned int]’
../../tests/../inc/catch2/../trompeloeil.hpp:3203:69:   required from ‘void trompeloeil::print_mismatch(std::ostream&, std::index_sequence<_Idx ...>, const std::tuple<_Tps ...>&, const std::tuple<_Elements ...>&) [with V = {D3D12_CPU_DESCRIPTOR_HANDLE, std::nullptr_t}; P = {D3D12_CPU_DESCRIPTOR_HANDLE&, unsigned int*&}; long long unsigned int ...I = {0, 1}; std::ostream = std::basic_ostream<char>; std::index_sequence<_Idx ...> = std::integer_sequence<long long unsigned int, 0, 1>]’
../../tests/../inc/catch2/../trompeloeil.hpp:3212:19:   required from ‘void trompeloeil::print_mismatch(std::ostream&, const std::tuple<_Tps ...>&, const std::tuple<_Elements ...>&) [with V = {D3D12_CPU_DESCRIPTOR_HANDLE, std::nullptr_t}; P = {D3D12_CPU_DESCRIPTOR_HANDLE&, unsigned int*&}; std::ostream = std::basic_ostream<char>]’
../../tests/../inc/catch2/../trompeloeil.hpp:4060:38:   required from ‘std::ostream& trompeloeil::call_matcher<Sig, Value>::report_mismatch(std::ostream&, trompeloeil::call_params_type_t<Sig>&) [with Sig = long int(D3D12_CPU_DESCRIPTOR_HANDLE, unsigned int*); Value = std::tuple<D3D12_CPU_DESCRIPTOR_HANDLE, std::nullptr_t>; std::ostream = std::basic_ostream<char>; trompeloeil::call_params_type_t<Sig> = std::tuple<D3D12_CPU_DESCRIPTOR_HANDLE&, unsigned int*&>]’
../../tests/../inc/catch2/../trompeloeil.hpp:4040:5:   required from here
../../tests/../inc/catch2/../trompeloeil.hpp:3143:42: error: no match for ‘operator==’ (operand types are ‘D3D12_CPU_DESCRIPTOR_HANDLE’ and ‘const D3D12_CPU_DESCRIPTOR_HANDLE’)
 3143 |     return ::trompeloeil::identity<U>(t) == u;
      |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~
In file included from /usr/share/mingw-w64/include/winnt.h:635,
                 from /usr/share/mingw-w64/include/minwindef.h:163,
                 from /usr/share/mingw-w64/include/windef.h:9,
                 from /usr/share/mingw-w64/include/windows.h:69,
                 from /usr/share/mingw-w64/include/rpc.h:16,
                 from /usr/share/mingw-w64/include/dxgi.h:7,
                 from ../../tests/../src/nvapi_private.h:15,
                 from ../../tests/../src/nvapi_d3d12.cpp:1,
                 from ../../tests/nvapi_d3d12.cpp:5:
/usr/share/mingw-w64/include/guiddef.h:180:15: note: candidate: ‘bool operator==(const GUID&, const GUID&)’
  180 | __inline bool operator== (REFGUID guidOne, REFGUID guidOther) { return !!IsEqualGUID (guidOne, guidOther); }
      |               ^~~~~~~~
/usr/share/mingw-w64/include/guiddef.h:180:35: note:   no known conversion for argument 1 from ‘D3D12_CPU_DESCRIPTOR_HANDLE’ to ‘const GUID&’ {aka ‘const _GUID&’}
  180 | __inline bool operator== (REFGUID guidOne, REFGUID guidOther) { return !!IsEqualGUID (guidOne, guidOther); }
`

yes, it works with `_`, issue is with `D3D12_CPU_DESCRIPTOR_HANDLE` .

Reply:
I am going to push with `_` for now. I hope we can merge with this?

Reply:
Yeah, fine to go with `_`. I'll take a loot after work, from a quick glance now everything looks pretty good. I'm certain we can merge this soon, likely today.

Reply:
@rochaudhari This did the trick: https://github.com/jp7677/dxvk-nvapi/commit/1f3a63bd6622310a6e58ea327952ed19ce5ebd74
Note that I changed some bits in the tests, For the "OK" tests I not just allowed, but required the corresponding `deviceExt` calls including how often they should be called (`TIMES(1)`). This makes the test much more useful since we now assert when the `deviceExt` aren't called.

Reply:
Interesting, `ptr` made a difference.!
Putting restrictions on number of times makes more sense.!